intro_to_ada.rst

```diff
@@ -1904,23 +1904,646 @@ Abstract data types
+| Any floating-point type | ``type T is digits <>;``                    |
++-------------------------+---------------------------------------------+
+
+Formal object declaration
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Formal objects are similar to subprogram parameters. Also, they can make
+use of formal types declared in the formal specification. For example:
+
+.. code-block:: ada
+
+    procedure Show_Formal_Object_Declaration is
+
+       generic
+          type T is private;
+          X : in out T;
+          --  X can be used in the Set procedure
+       procedure Set (E : in T);
+
+       procedure Set (E : in T) is null;
+
+    begin
+       null;
+    end Show_Formal_Object_Declaration;
+
+Formal objects can be either just input parameters or use the ``in out``
+mode.
+
+Generic body definition
+~~~~~~~~~~~~~~~~~~~~~~~
+
+For the body declaration of a generic subprogram or package, we don't
+repeat the ``generic`` keyword: we simply start with the actual
+declaration and make use of the generic types and objects that we
+declared. For example:
+
+.. code-block:: ada
+
+    procedure Show_Generic_Body_Definition is
+
+       generic
+          type T is private;
+          X : in out T;
+       procedure Set (E : T);
+
+       procedure Set (E : T) is
+       --  Body definition: "generic" keyword is not used
+       begin
+          X := E;
+       end Set;
+    begin
+       null;
+    end Show_Generic_Body_Definition;
+
 Generic instantiation
----------------------
+~~~~~~~~~~~~~~~~~~~~~
 
-Formal types
-------------
+Generic subprograms or packages cannot be used directly. Instead, they
+need to be instantiated. The instantiation is done by using the ``new``
+keyword, as illustrated in the following example:
 
-Formal objects
---------------
+.. code-block:: ada
+
+    with Ada.Text_IO;
+    use  Ada.Text_IO;
+
+    procedure Show_Generic_Instantiation is
+
+       generic
+          type T is private;
+          X : in out T;
+          --  X can be used in the Set procedure
+       procedure Set (E : T);
+
+       procedure Set (E : T) is
+       begin
+          X := E;
+       end Set;
+
+       Main    : Integer := 0;
+       Current : Integer;
+
+       procedure Set_Main is new Set (T => Integer,
+                                      X => Main);
+       --  Here, we map the formal parameters to actual types and objects.
+       --
+       --  The same approach can be used to instantiate functions or
+       --  packages, e.g.:
+       --  function Get_Main is new ...
+       --  package Integer_Queue is new ...
+
+    begin
+       Current := 10;
+
+       Set_Main (Current);
+       Put_Line ("Value of Main is " & Integer'Image (Main));
+    end Show_Generic_Instantiation;
+
+In the example above, we instantiate the procedure ``Set`` by mapping the
+formal parameters ``T`` and ``X`` to actual existing elements: the
+``Integer`` type and the ``Main`` variable.
 
 Formal subprograms
```

Comment:
I think before that you should talk about generic packages. I'm talking about plain generic packages, like so:

```ada
generic
    type T is private
package Smart_Pointer is
end Smart_Pointer;
```
Because this is the most important construct with generics (basis of ADTs).

I'm not talking about passing generic packages to other generic packages, which we determined last week was advanced :)